File: backend-app/src/main/java/com/backend_app/backend_app/service/PrestamoService.java
package com.backend_app.backend_app.service;

import com.backend_app.backend_app.domain.Estado;
import com.backend_app.backend_app.domain.Prestamo;
import com.backend_app.backend_app.dto.EstadoDTO;
import com.backend_app.backend_app.dto.PrestamoDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Optional;

public interface PrestamoService {

    List<PrestamoDTO> listadoPrestamos();

    Optional<PrestamoDTO> buscarPrestamoPorId(long idPrestamo);

    PrestamoDTO convertirADTO(Prestamo prestamo);

    void insertarPrestamo(PrestamoDTO prestamoDTO);

    void actualizarPrestamo(PrestamoDTO prestamoDTO);

    void eliminarPrestamo(PrestamoDTO prestamoDTO);

    Page<PrestamoDTO> findAll(Pageable pageable);

    List<EstadoDTO> obtenerEstadosPrestamos();

    List<PrestamoDTO> listadoPrestamosCompletos();

    List<EstadoDTO> prestamosPorEstado();


}

File: frontend-app/src/app/services/prestamo.service.ts
import { HttpClient } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { Prestamo } from "../domain/Prestamo";
import { Estado } from "../domain/Estado";


@Injectable(
    {
        providedIn: 'root'
    }
)
export class PrestamoService{

    private url: string = 'http://localhost:8080/api/prestamos';

    constructor(private http: HttpClient){

    }

    listadoPrestamos(): Observable<Prestamo[]>{
        return this.http.get<Prestamo[]>(this.url);
    }

    buscarPrestamoPorId(id: number): Observable<Prestamo>{
        return this.http.get<Prestamo>(`${this.url}/${id}`);
    }

    guardarPrestamo(prestamo: Prestamo){
        return this.http.post('http://localhost:8080/api/prestamos/guardar', prestamo);
    }
    
    editarPrestamo(prestamo: Prestamo) {
        return this.http.put(`${this.url}/actualizar`, prestamo);
    }
    
    eliminarPrestamo(idPrestamo: number){
        return this.http.delete(`${this.url}/${idPrestamo}`);
    }

    listadoPrestamosPaginacion(page: number): Observable<any>{
        return this.http.get<any[]>(`${this.url}/page/${page}`);
    }   

    obtenerEstadoPrestamos(): Observable<Estado[]>{
        return this.http.get<Estado[]>(`${this.url}/estado`);
    }

    listadoPrestamosCompletos(): Observable<Prestamo[]>{
        return this.http.get<Prestamo[]>(`${this.url}/completos`)
    }

    prestamosPorEstado(): Observable<Estado[]> {
        return this.http.get<Estado[]>(`${this.url}/por-estado`);
    }

}